src/ray/core_worker/core_worker.cc

```diff
@@ -1413,17 +1418,16 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
   task_queue_length_ -= 1;
   num_executed_tasks_ += 1;
 
-  if (resource_ids != nullptr) {
-    resource_ids_ = resource_ids;
-  }
-
   if (!options_.is_local_mode) {
     worker_context_.SetCurrentTask(task_spec);
     SetCurrentTaskId(task_spec.TaskId());
   }
   {
     absl::MutexLock lock(&mutex_);
     current_task_ = task_spec;
+    if (resource_ids) {
+      resource_ids_ = resource_ids;
+    }
```

Comment:
I don't think this actually fixes it. The assign should be safe, the problem is we are iterating over the held object while it is deleted out from under us.

Reply:
I think it's fine as long as the iteration happens under the same lock.

Reply:
Well, maybe the lock is needed for the copy to be safe (not sure).

Reply:
Ah got it.